File: src/test/kotlin/io/miiken/intellijcontrolserver/fixtures/kotlin/SampleClass.kt
package io.miiken.intellijcontrolserver.fixtures.kotlin

import org.slf4j.LoggerFactory
import io.micrometer.core.annotation.Timed

/**
 * Sample Kotlin class for testing refactoring operations.
 * 
 * This class includes test cases for renaming:
 * - Methods
 * - Fields
 * - Parameters
 * - Local variables
 * - Classes
 */
class SampleCalculator {
    
    // RENAME THIS - Test renaming this field
    private var oldFieldName: Int = 10

    private val logger = LoggerFactory.getLogger(javaClass)
    
    // RENAME THIS - Test renaming this method
    @Timed("metrics.timer.oldMethodName")
    fun oldMethodName(oldParameterName: Int): Int {
        logger.info("oldMethodName: $oldParameterName")
        // RENAME THIS - Test renaming this local variable
        val oldVariableName = oldParameterName * 2
        return oldVariableName + oldFieldName
    }
    
    /**
     * Calculate discount based on total amount.
     * Calls oldMethodName to apply bonus points.
     */
    fun calculateDiscount(total: Double, membershipLevel: String): Double {
        val baseDiscount = when (membershipLevel) {
            "gold" -> 0.20
            "silver" -> 0.10
            "bronze" -> 0.05
            else -> 0.0
        }
        
        val volumeDiscount = if (total > 100.0) 0.05 else 0.0
        
        // Call oldMethodName to calculate bonus points
        val bonusPoints = oldMethodName(total.toInt())
        oldFieldName += bonusPoints
        
        return (baseDiscount + volumeDiscount).coerceAtMost(0.30)
    }
    
    /**
     * Process a purchase.
     * Calls calculateDiscount and oldMethodName.
     */
    fun processPurchase(amount: Double, level: String): Double {
        val discount = calculateDiscount(amount, level)
        val finalAmount = amount * (1.0 - discount)
        
        // Update oldFieldName through oldMethodName
        oldMethodName(amount.toInt())
        
        return finalAmount
    }
}

File: src/test/resources/testdata/typescript/SampleClass.ts
/**
 * Sample TypeScript class for testing refactoring operations.
 * 
 * This class includes:
 * - A method to rename (marked with RENAME THIS)
 * - A method with extractable code (marked with EXTRACT START/END)
 */
class SampleCalculator {
    
    // RENAME THIS - Test renaming this method
    oldMethodName(value: number): number {
        return value * 2;
    }
    
    /**
     * Process an order with multiple items.
     * Contains code that should be extracted into a separate method.
     */
    processOrder(userId: string, items: string[]): void {
        console.log(`Processing order for user: ${userId}`);
        
        // EXTRACT START (lines 23-28)
        let total: number = 0.0;
        for (const item of items) {
            const price = this.getPrice(item);
            total += price;
        }
        const tax: number = total * 0.1;
        // EXTRACT END
        
        const finalTotal = total + tax;
        console.log(`Total for ${userId}: $${finalTotal}`);
        this.saveOrder(userId, finalTotal);
    }
    
    private getPrice(item: string): number {
        switch (item) {
            case 'book':
                return 15.99;
            case 'pen':
                return 2.50;
            case 'notebook':
                return 5.99;
            default:
                return 10.0;
        }
    }
    
    private saveOrder(userId: string, total: number): void {
        console.log(`Order saved: ${userId} -> $${total}`);
    }
    
    /**
     * Calculate discount based on total amount.
     * This demonstrates TypeScript features with type annotations.
     */
    calculateDiscount(total: number, membershipLevel: string): number {
        let baseDiscount: number;
        switch (membershipLevel) {
            case 'gold':
                baseDiscount = 0.20;
                break;
            case 'silver':
                baseDiscount = 0.10;
                break;
            case 'bronze':
                baseDiscount = 0.05;
                break;
            default:
                baseDiscount = 0.0;
        }
        
        const volumeDiscount = total > 100.0 ? 0.05 : 0.0;
        return Math.min(baseDiscount + volumeDiscount, 0.30);
    }
}

export default SampleCalculator;
